### ISSUE TITLE
A question about exact model architecture !

### ISSUE THREAD

**Issue by khahn0213** [aosokin/os2d#2]
Hello. First of all, I'm very appreciative of your great work!

I am now trying to apply your os2d-v2-trained model to my custom data.
However, I cannot improve evaluate result since the model is not able to be trained because of
no-annotation with my custom data.
So during tagging my custom data, I'm planning to insert other modules to your os2d-v2-trained model, but I think I cannot get access to the exact architecture of the model, for example changing anchors but I cannot find where they are.

I'm very looking forward to your reply.
Thank you.

**Comment by aosokin** [comment]
Hi @khahn0213 , thanks for the kind words.
Most of the details about the model architecture are in [the modeling folder](https://github.com/aosokin/os2d/tree/master/os2d/modeling).

Specifically, anchor boxes are created in the [box_coder.py](https://github.com/aosokin/os2d/blob/master/os2d/modeling/box_coder.py) file:
https://github.com/aosokin/os2d/blob/a1a503164bb760f75bae42d26899fc217cf172d9/os2d/modeling/box_coder.py#L191
This function calls 
https://github.com/aosokin/os2d/blob/a1a503164bb760f75bae42d26899fc217cf172d9/os2d/modeling/box_coder.py#L17
to generate a grid of boxes.

**Comment by khahn0213** [comment]
Thanks for the kind and immediate comment!
I am checking the modeling folder and the box_coder.py file now as your advice! 

During testing on my custom dataset, I have another question!
I've found that you update the margin part of  OS2D loss. 
After following your update, I've found that the bounding box is less accurate than before (I mean not tighter than before), but anyhow get the region of the object more than before even the object is so small compared to resolution of the image.

I think this is better than before, but I want to let the bounding box much tighter to every object even it is tiny. I've found that I can resize or add zero-padding to tiny or big images to improve the result but this is not an essential solution....

I've also tried changing max_detection, threshold for IoU and threshold_score to get neat result of bounding box, but this only gives me many tight bbox on one object, not spread bbox to each object.

Would you give me how can I solve those problems?

And again, Thanks for the amazing work!
(I find your awesome work through twitter though!)

**Comment by aosokin** [comment]
> I've found that you update the margin part of OS2D loss.
> After following your update, I've found that the bounding box is less accurate than before (I mean not tighter than before), but anyhow get the region of the object more than before even the object is so small compared to resolution of the image.

I'm sorry, but I do not understand which update you are talking about. Could you give more details?

> I think this is better than before, but I want to let the bounding box much tighter to every object even it is tiny. I've found that I can resize or add zero-padding to tiny or big images to improve the result but this is not an essential solution....

When you evaluate at one scale the method detects objects, which are around a specific size: 240x240 pixels (comes from the working size of TransformNet, which is 15x15 in the feature map, and the effective receptive field of the feature extractor). If you want to detect much smaller objects the only currently supported way is to add images at a larger resolution.

> I've also tried changing max_detection, threshold for IoU and threshold_score to get neat result of bounding box, but this only gives me many tight bbox on one object, not spread bbox to each object.

These parameters work at the NMS stage (postprocessing), but not on the detector itself.

**Comment by khahn0213** [comment]
> I'm sorry, but I do not understand which update you are talking about. Could you give more details?

Oh I'm sorry. What I said about your update is the change of the margin part of OS2D loss as I put the link of your commit. 

https://github.com/aosokin/os2d/commit/a1a503164bb760f75bae42d26899fc217cf172d9

> When you evaluate at one scale the method detects objects, which are around a specific size: 
> 240x240 pixels (comes from the working size of TransformNet, which is 15x15 in the feature map,
> and the effective receptive field of the feature extractor). If you want to detect much smaller 
> objects the only currently supported way is to add images at a larger resolution.

Ok. I understand it. Thanks for your reply!

> These parameters work at the NMS stage (postprocessing), but not on the detector itself.

Yes but when I let cfg.eval.nms_score_threshold >=0 in this part,
https://github.com/aosokin/os2d/blob/a1a503164bb760f75bae42d26899fc217cf172d9/os2d/config.py#L197-L200

I get 'out of range' error at this point.
https://github.com/aosokin/os2d/blob/a1a503164bb760f75bae42d26899fc217cf172d9/os2d/structures/bounding_box.py#L402-L403

**Comment by aosokin** [comment]
> a1a5031

The key idea of this commit was to get rid of confusing
```output_recognition = (matches_summed - 1.0) / 2.0```
in https://github.com/aosokin/os2d/blob/a1a503164bb760f75bae42d26899fc217cf172d9/os2d/modeling/head.py#L308
This was causing several different versions of detection scores in the code. I think I've modified other parts accordingly: the way margins are applied in the loss (now it matches the way it's written in the paper), the margin constants in the code everywhere. If something went wrong please give me an example.

> I get 'out of range' error at this point.

This most likely means that you passed an empty list to `cat_boxlist`. I can try to help if you give a minimal reproducing example

**Comment by khahn0213** [comment]
> If something went wrong please give me an example.

> This most likely means that you passed an empty list to cat_boxlist. I can try to help if you give a 
> minimal reproducing example

Sorry. I made some mistakes while I change the origin code over and over again. 
I've found that for my custom dataset, 
cfg.eval.nms_iou_threshold = 0.2 
cfg.eval.nms_score_threshold = float("-inf") ~ -0.3 
works well :-) .

I really appreciate for your kind help!
Your work and your code drive me into a deeper study! 
(Actually, I've spent half of the day to debug line to line to check every variable and its dimension...)
